### PR TITLE
feature add singleflight cache

### PR DESCRIPTION
**Pull request by Stone-afk**
add singleflight cache, test case not supplemented

### CODE REVIEW

**Comment by flycash** [comment]
please  rebase to origin/develop and fix the conflicts

**Comment by Stone-afk** [comment]
目前针对 ssdb 的并发测试，不管设置的 value 是啥最终都会会持续返回 “1”

**Comment by Stone-afk** [comment]
`
func testSingleflightCacheConcurrencyGet(t *testing.T, bm cache.Cache) {
	key, value := "key3", "value3"
	db := &MockOrm{keysMap: map[string]int{key: 1}, kvs: map[string]any{key: value}}
	c, err := cache.NewSingleflightCache(bm, 10*time.Second,
		func(ctx context.Context, key string) (any, error) {
			val, er := db.Load(key)
			if er != nil {
				return nil, er
			}
			return val, nil
		})
	assert.Nil(t, err)

	var wg sync.WaitGroup
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go func() {
			defer wg.Done()
			val, err := c.Get(context.Background(), key)
			if err != nil {
				t.Error(err)
			}
			assert.Equal(t, value, val)
		}()
	}
	wg.Wait()
}
`

**Comment by Stone-afk** [comment]
```
func testSingleflightCacheConcurrencyGet(t *testing.T, bm cache.Cache) {
	key, value := "key3", "value3"
	db := &MockOrm{keysMap: map[string]int{key: 1}, kvs: map[string]any{key: value}}
	c, err := cache.NewSingleflightCache(bm, 10*time.Second,
		func(ctx context.Context, key string) (any, error) {
			val, er := db.Load(key)
			if er != nil {
				return nil, er
			}
			return val, nil
		})
	assert.Nil(t, err)


	var wg sync.WaitGroup
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go func() {
			defer wg.Done()
			val, err := c.Get(context.Background(), key)
			if err != nil {
				t.Error(err)
			}
			assert.Equal(t, value, val)
		}()
	}
	wg.Wait()
}

```

**Review by flycash — CHANGES_REQUESTED** [review]
remove other tests. Just keep singleflight_test

File: client/cache/singleflight.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"golang.org/x/sync/singleflight"
+)
+
+// SingleflightCache
+// This is a very simple decorator mode
+type SingleflightCache struct {
+	Cache
+	group      *singleflight.Group
+	Expiration time.Duration
+	LoadFunc   func(ctx context.Context, key string) (any, error)
```
Comment (flycash) [inline client/cache/singleflight.go:32]: make them private

File: client/cache/singleflight.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		expiration: expiration,
+		loadFunc:   loadFunc,
+	}, nil
+}
+
+// Get In the Get method, single flight is used to load data and write back the cache.
+func (s *SingleflightCache) Get(ctx context.Context, key string) (any, error) {
+	val, err := s.Cache.Get(ctx, key)
+	if val == nil || err != nil {
+		val, err, _ = s.group.Do(key, func() (interface{}, error) {
```
Comment (flycash) [inline client/cache/singleflight.go:53]: I think you need to delete the key after loading data. So when the key is expired again, goroutine will load data one more time

**Review by flycash — APPROVED** [review]

**Pull request merged**